error.go

```diff
@@ -72,22 +65,271 @@ func collectMessages(data interface{}, messages []string) []string {
+type RetryError struct {
+	APIError
+}
+
+// ServiceUnavailableError is raised when the API returns a 503 status code.
+type ServiceUnavailableError struct {
+	APIError
+}
+
+// SSLError is raised when there is an issue with the SSL certificate.
+type SSLError struct {
+	APIError
+}
+
+// TimeoutError is raised when the API returns a 408 status code.
+type TimeoutError struct {
+	APIError
+}
+
+// UnauthorizedError is raised when the API returns a 401 status code.
+type UnauthorizedError struct {
+	APIError
+}
+
+// ForbiddenError is raised when the API returns a 403 status code.
+type ForbiddenError struct {
+	APIError
+}
+
+// UnknownHttpError is raised when the API returns an unrecognized status code.
+type UnknownHttpError struct {
+	APIError
+}
+
+// BuildErrorFromResponse returns an APIError-based object based on the HTTP response.
+// Do not pass a non-error response to this function.
+func BuildErrorFromResponse(response *http.Response) error {
+	// build the base APIError object from the response
+	apiError := &APIError{
+		StatusCode: response.StatusCode,
+	}
+
+	// deserialize the response body into a temporary object
+	buf, _ := ioutil.ReadAll(response.Body)
+	tmpError := &struct {
+		Error *Error `json:"error,omitempty"`
+	}{}
+
+	if json.Unmarshal(buf, &tmpError) == nil {
+		// extract the details from the temporary object (top-level Error class) and store them in the APIError object
+		apiError.Message = tmpError.Error.Message.(string)
+		apiError.Code = tmpError.Error.Code
+		apiError.Errors = tmpError.Error.Errors
+	} else {
+		// could not extract error details from the API response (or API did not return data, i.e. 1xx, 3xx or 5xx)
+		if response.Status == "" {
+			response.Status = ApiDidNotReturnErrorDetails
+		}
+		apiError.Message = response.Status
+		apiError.Code = ApiErrorDetailsParsingError
+		apiError.Errors = []*Error{}
+	}
+
+	// return the appropriate error type based on the status code
+	switch response.StatusCode {
+	case 0:
+		return &ConnectionError{APIError: *apiError}
+	case 100, 101, 102, 103:
+		return &RetryError{APIError: *apiError}
+	case 300, 301, 302, 303, 304, 305, 306, 307, 308:
+		return &RedirectError{APIError: *apiError}
+	case 400:
+		return &BadRequestError{APIError: *apiError}
+	case 401:
+		return &UnauthorizedError{APIError: *apiError}
+	case 402:
+		return &PaymentError{APIError: *apiError}
+	case 403:
+		return &ForbiddenError{APIError: *apiError}
+	case 404:
+		return &NotFoundError{APIError: *apiError}
+	case 405:
+		return &MethodNotAllowedError{APIError: *apiError}
+	case 407:
+		return &ProxyError{APIError: *apiError}
+	case 408:
+		return &TimeoutError{APIError: *apiError}
+	case 422:
+		return &InvalidRequestError{APIError: *apiError}
+	case 429:
+		return &RateLimitError{APIError: *apiError}
+	case 500:
+		return &InternalServerError{APIError: *apiError}
+	case 503:
+		return &ServiceUnavailableError{APIError: *apiError}
+	case 504:
```

Comment:
```suggestion
	case 502, 504:
```
Our API returns 502, let's add it so this will not fall to `UnknownHttpError`